Assets-only workers broken by assets validation

### Describe the bug

The validation in #13844 broke assets-only workers, which do _not_ specify `binding = "ASSETS"`. If you include that, you get the error "[ERROR] Cannot use assets with a binding in an assets-only Worker." from wrangler. However, without a binding, you get "Error: You must specify the `assets.binding` key in /opt/buildhome/repo/wrangler.toml. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding"

### Reproduction

Create a wrangler.toml including `binding = "ASSETS"` under `[assets]` but no `main` key.

### Logs

```Shell

```

### System Info

```Shell
System:
    OS: macOS 15.5
    CPU: (10) arm64 Apple M1 Max
    Memory: 1.31 GB / 64.00 GB
    Shell: 5.9 - /run/current-system/sw/bin/zsh
  Binaries:
    Node: 22.16.0 - /nix/store/abxz7wpw5kwaijzzmf8rl4ga981vk8pc-nodejs-22.16.0/bin/node
    npm: 10.9.2 - /nix/store/abxz7wpw5kwaijzzmf8rl4ga981vk8pc-nodejs-22.16.0/bin/npm
    pnpm: 9.15.0 - /nix/store/6ay4p54n66jj9npcn362djdhbb5cwi7a-corepack-nodejs-22.16.0/bin/pnpm
  Browsers:
    Chrome: 138.0.7204.158
    Safari: 18.5
  npmPackages:
    @sveltejs/adapter-cloudflare: ^7.0.5 => 7.0.5
    @sveltejs/enhanced-img: ^0.6.1 => 0.6.1
    @sveltejs/kit: ^2.25.1 => 2.25.1
    @sveltejs/vite-plugin-svelte: ^6.1.0 => 6.1.0
    svelte: ^5.36.8 => 5.36.8
    vite: ^7.0.5 => 7.0.5
```

### Severity

blocking an upgrade

### Additional Information

_No response_

--- a/.changeset/warm-rats-say.md
+++ b/.changeset/warm-rats-say.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/adapter-cloudflare': patch
+---
+
+fix: support assets-only workers in wrangler validation

--- a/packages/adapter-cloudflare/index.js
+++ b/packages/adapter-cloudflare/index.js
@@ -2,7 +2,7 @@ import { VERSION } from '@sveltejs/kit';
 import { copyFileSync, existsSync, writeFileSync } from 'node:fs';
 import path from 'node:path';
 import { fileURLToPath } from 'node:url';
-import { is_building_for_cloudflare_pages } from './utils.js';
+import { is_building_for_cloudflare_pages, validate_worker_settings } from './utils.js';
 import { getPlatformProxy, unstable_readConfig } from 'wrangler';
 
 const name = '@sveltejs/adapter-cloudflare';
@@ -31,7 +31,7 @@ export default function (options = {}) {
 				);
 			}
 
-			const wrangler_config = validate_config(options.config);
+			const wrangler_config = validate_wrangler_config(options.config);
 			const building_for_cloudflare_pages = is_building_for_cloudflare_pages(wrangler_config);
 
 			let dest = builder.getBuildDirectory('cloudflare');
@@ -286,33 +286,12 @@ _redirects
  * @param {string} config_file
  * @returns {import('wrangler').Unstable_Config}
  */
-function validate_config(config_file = undefined) {
+function validate_wrangler_config(config_file = undefined) {
 	const wrangler_config = unstable_readConfig({ config: config_file });
 
-	// we don't support workers sites
-	if (wrangler_config.site) {
-		throw new Error(
-			`You must remove all \`site\` keys in ${wrangler_config.configPath}. Consult https://svelte.dev/docs/kit/adapter-cloudflare#Migrating-from-Workers-Sites-to-Workers-Static-Assets`
-		);
-	}
-
-	if (is_building_for_cloudflare_pages(wrangler_config)) {
-		return wrangler_config;
-	}
-
-	// probably deploying to Cloudflare Workers
-	if (wrangler_config.main || wrangler_config.assets) {
-		if (!wrangler_config.assets?.directory) {
-			throw new Error(
-				`You must specify the \`assets.directory\` key in ${wrangler_config.configPath}. Consult https://developers.cloudflare.com/workers/static-assets/binding/#directory`
-			);
-		}
-
-		if (!wrangler_config.assets?.binding && wrangler_config.main) {
-			throw new Error(
-				`You must specify the \`assets.binding\` key in ${wrangler_config.configPath}. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding`
-			);
-		}
+	if (!is_building_for_cloudflare_pages(wrangler_config)) {
+		// probably deploying to Cloudflare Workers
+		validate_worker_settings(wrangler_config);
 	}
 
 	return wrangler_config;

--- a/packages/adapter-cloudflare/index.js
+++ b/packages/adapter-cloudflare/index.js
@@ -308,9 +308,15 @@ function validate_config(config_file = undefined) {
 			);
 		}
 
-		if (!wrangler_config.assets?.binding) {
+		if (!wrangler_config.assets?.binding && wrangler_config.main) {
 			throw new Error(
 				`You must specify the \`assets.binding\` key in ${wrangler_config.configPath}. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding`
+			);
+		}
+
+		if (wrangler_config.assets?.binding && !wrangler_config.main) {
+			throw new Error(
+				`You must remove the \`assets.binding\` key in ${wrangler_config.configPath} for an assets-only Worker, which has no \`main\` key.`
 			);
 		}
 	}

--- a/packages/adapter-cloudflare/index.js
+++ b/packages/adapter-cloudflare/index.js
@@ -313,12 +313,6 @@ function validate_config(config_file = undefined) {
 				`You must specify the \`assets.binding\` key in ${wrangler_config.configPath}. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding`
 			);
 		}
-
-		if (wrangler_config.assets?.binding && !wrangler_config.main) {
-			throw new Error(
-				`You must remove the \`assets.binding\` key in ${wrangler_config.configPath} for an assets-only Worker, which has no \`main\` key.`
-			);
-		}
 	}
 
 	return wrangler_config;

--- a/packages/adapter-cloudflare/utils.js
+++ b/packages/adapter-cloudflare/utils.js
@@ -15,3 +15,37 @@ export function is_building_for_cloudflare_pages(wrangler_config) {
 
 	return true;
 }
+
+/**
+ * @param {import('wrangler').Unstable_Config} wrangler_config
+ */
+export function validate_worker_settings(wrangler_config) {
+	// we don't support workers sites
+	if (wrangler_config.site) {
+		throw new Error(
+			`You must remove all \`site\` keys in ${wrangler_config.configPath}. Consult https://svelte.dev/docs/kit/adapter-cloudflare#Migrating-from-Workers-Sites-to-Workers-Static-Assets`
+		);
+	}
+
+	// we need the `assets.directory` key so that the static assets are deployed
+	if ((wrangler_config.main || wrangler_config.assets) && !wrangler_config.assets?.directory) {
+		throw new Error(
+			`You must specify the \`assets.directory\` key in ${wrangler_config.configPath}. Consult https://developers.cloudflare.com/workers/static-assets/binding/#directory`
+		);
+	}
+
+	// we need the `assets.binding` key so that the Worker can access the static assets
+	if (wrangler_config.main && !wrangler_config.assets?.binding) {
+		throw new Error(
+			`You must specify the \`assets.binding\` key in ${wrangler_config.configPath} before deploying your Worker. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding`
+		);
+	}
+
+	// the user might have forgot the `main` key or should remove the `assets.binding`
+	// key to deploy static assets without a Worker
+	if (!wrangler_config.main && wrangler_config.assets?.binding) {
+		throw new Error(
+			`You must set the \`main\` key in ${wrangler_config.configPath} if you want to deploy a Worker alongside your static assets or remove the \`assets.binding\` key if you only want to deploy static assets.`
+		);
+	}
+}

--- a/packages/adapter-cloudflare/utils.spec.js
+++ b/packages/adapter-cloudflare/utils.spec.js
@@ -1,5 +1,5 @@
 import { describe, test, vi, expect } from 'vitest';
-import { is_building_for_cloudflare_pages } from './utils.js';
+import { is_building_for_cloudflare_pages, validate_worker_settings } from './utils.js';
 
 describe('detects Cloudflare Pages project', () => {
 	test('by default', () => {
@@ -57,11 +57,88 @@ describe('detects Cloudflare Workers project', () => {
 				/** @type {import('wrangler').Unstable_Config} */ ({
 					configPath: 'wrangler.jsonc',
 					assets: {
+						directory: 'dist/assets'
+					}
+				})
+			)
+		).toBe(false);
+	});
+});
+
+describe('validates Wrangler config', () => {
+	test('Worker and static assets', () => {
+		expect(() =>
+			validate_worker_settings(
+				/** @type {import('wrangler').Unstable_Config} */ ({
+					configPath: 'wrangler.jsonc',
+					main: 'dist/index.js',
+					assets: {
 						directory: 'dist/assets',
 						binding: 'ASSETS'
 					}
 				})
 			)
-		).toBe(false);
+		).not.toThrow();
+	});
+
+	test('static assets only', () => {
+		expect(() =>
+			validate_worker_settings(
+				/** @type {import('wrangler').Unstable_Config} */ ({
+					configPath: 'wrangler.jsonc',
+					assets: {
+						directory: 'dist/assets'
+					}
+				})
+			)
+		).not.toThrow();
+	});
+
+	test('missing `assets.directory` key', () => {
+		expect(() =>
+			validate_worker_settings(
+				/** @type {import('wrangler').Unstable_Config} */ ({
+					configPath: 'wrangler.jsonc',
+					main: 'dist/index.js',
+					assets: {
+						binding: 'ASSETS'
+					}
+				})
+			)
+		).toThrow(
+			`You must specify the \`assets.directory\` key in wrangler.jsonc. Consult https://developers.cloudflare.com/workers/static-assets/binding/#directory`
+		);
+	});
+
+	test('missing `assets.binding` key', () => {
+		expect(() =>
+			validate_worker_settings(
+				/** @type {import('wrangler').Unstable_Config} */ ({
+					configPath: 'wrangler.jsonc',
+					main: 'dist/index.js',
+					assets: {
+						directory: 'dist/assets'
+					}
+				})
+			)
+		).toThrow(
+			`You must specify the \`assets.binding\` key in wrangler.jsonc before deploying your Worker. Consult https://developers.cloudflare.com/workers/static-assets/binding/#binding`
+		);
+	});
+
+	test('missing `main` key or should remove `assets.binding` key', () => {
+		expect(() =>
+			validate_worker_settings(
+				/** @type {import('wrangler').Unstable_Config} */ ({
+					configPath: 'wrangler.jsonc',
+					assets: {
+						directory: 'dist/assets',
+						binding: 'ASSETS'
+					}
+				})
+			)
+		).toThrow(
+			`You must set the \`main\` key in wrangler.jsonc if you want to deploy a Worker alongside your static assets or remove the \`assets.binding\` key if you only want to deploy static assets.`
+		);
 	});
 });

